README.md
go-unidecode
==============
[![Build Status](https://app.travis-ci.com/lemontreeran/go-unidecode.svg?branch=master)](https://travis-ci.com/github/lemontreeran/go-unidecode)
[![Coverage Status](https://coveralls.io/repos/lemontreeran/go-unidecode/badge.svg?branch=master)](https://coveralls.io/r/lemontreeran/go-unidecode?branch=master)
[![Go Report Card](https://goreportcard.com/badge/github.com/lemontreeran/go-unidecode)](https://goreportcard.com/report/github.com/lemontreeran/go-unidecode)
[![GoDoc](https://godoc.org/github.com/lemontreeran/go-unidecode?status.svg)](https://godoc.org/github.com/lemontreeran/go-unidecode)

ASCII transliterations of Unicode text.


Installation
------------

```
go get -u github.com/lemontreeran/go-unidecode
```

install CLI tool:

```
go get -u github.com/lemontreeran/go-unidecode/unidecode

$ unidecode 北京kožušček
Bei Jing kozuscek
```


Usage
------

```go
package main
import (
	"fmt"
	"github.com/lemontreeran/go-unidecode"
)
func main() {
	s = "abc"
	fmt.Println(unidecode.Unidecode(s))
	// Output: abc
	s = "北京"
	fmt.Println(unidecode.Unidecode(s))
	// Output: Bei Jing
	s = "30 𝗄𝗆/𝗁"
	fmt.Println(unidecode.Unidecode(s))
	// Output: 30 km/h
	s = "kožušček"
	fmt.Println(unidecode.Unidecode(s))
	// Output: kozuscek
}
```


CHANGELOG.md
# Changelog


## 0.1.1 (2021-10-01)

* Initial Release


example_test.go
package unidecode_test

import (
	"fmt"

	"github.com/lemontreeran/go-unidecode"
)

func ExampleUnidecode() {
	s := "北京kožušček"
	fmt.Println(unidecode.Unidecode(s))
	// Output: Bei Jing kozuscek
}

func ExampleUnidecode_ascii() {
	s := "abc"
	fmt.Println(unidecode.Unidecode(s))
	// Output: abc
}


unidecode_test.go
package unidecode

import (
	"testing"
	"unicode"
)

type testCase struct {
	input, expect string
}

func testUnidecode(t *testing.T, input, expect string) {
	ret := Unidecode(input)
	check(t, ret, expect)
}

func check(t *testing.T, ret, expect string) {
	if ret != expect {
		t.Errorf("Expected %s, got %s", expect, ret)
	}
}

func TestVersion(t *testing.T) {
	ret := Version()
	expect := "0.1.1"
	check(t, ret, expect)
}

func TestUnidecodeASCII(t *testing.T) {
	for n := 0; n < unicode.MaxASCII; n++ {
		expect := string(rune(n))
		testUnidecode(t, string(rune(n)), expect)
	}
}

func TestUnidecode(t *testing.T) {
	cases := []testCase{
		{"", ""},
		{"abc", "abc"},
		{"北京", "Bei Jing "},
		{"abc北京", "abcBei Jing "},
		{"ネオアームストロングサイクロンジェットアームストロング砲", "neoamusutorongusaikuronzietsutoamusutoronguPao "},
		{"30 𝗄𝗆/𝗁", "30 km/h"},
		{"kožušček", "kozuscek"},
		{"ⓐⒶ⑳⒇⒛⓴⓾⓿", "aA20(20)20.20100"},
		{"Hello, World!", "Hello, World!"},
		{`\n`, `\n`},
		{`北京abc\n`, `Bei Jing abc\n`},
		{`'"\r\n`, `'"\r\n`},
		{"ČŽŠčžš", "CZSczs"},
		{"ア", "a"},
		{"α", "a"},
		{"a", "a"},
		{"ch\u00e2teau", "chateau"},
		{"vi\u00f1edos", "vinedos"},
		{"Efﬁcient", "Efficient"},
		{"příliš žluťoučký kůň pěl ďábelské ódy", "prilis zlutoucky kun pel dabelske ody"},
		{"PŘÍLIŠ ŽLUŤOUČKÝ KŮŇ PĚL ĎÁBELSKÉ ÓDY", "PRILIS ZLUTOUCKY KUN PEL DABELSKE ODY"},
		{"\ua500", ""},
		{"\u1eff", ""},
		// Fix wrong int to string conversions
		{string(rune(0xfffff)), ""},
		{"\U0001d5a0", "A"},
		{"\U0001d5c4\U0001d5c6/\U0001d5c1", "km/h"},
		{"\u2124\U0001d552\U0001d55c\U0001d552\U0001d55b \U0001d526\U0001d52a\U0001d51e \U0001d4e4\U0001d4f7\U0001d4f2\U0001d4ec\U0001d4f8\U0001d4ed\U0001d4ee \U0001d4c8\U0001d4c5\u212f\U0001d4b8\U0001d4be\U0001d4bb\U0001d4be\U0001d4c0\U0001d4b6\U0001d4b8\U0001d4be\U0001d4bf\u212f \U0001d59f\U0001d586 \U0001d631\U0001d62a\U0001d634\U0001d622\U0001d637\U0001d626?!", "Zakaj ima Unicode specifikacije za pisave?!"},
	}
	for _, c := range cases {
		testUnidecode(t, c.input, c.expect)
	}
}

func TestUnidecodeConverterA(t *testing.T) {
	v := "THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG 1234567890"
	cases := []string{
		"\uff34\uff28\uff25 \uff31\uff35\uff29\uff23\uff2b \uff22\uff32\uff2f\uff37\uff2e \uff26\uff2f\uff38 \uff2a\uff35\uff2d\uff30\uff33 \uff2f\uff36\uff25\uff32 \uff34\uff28\uff25 \uff2c\uff21\uff3a\uff39 \uff24\uff2f\uff27 \uff11\uff12\uff13\uff14\uff15\uff16\uff17\uff18\uff19\uff10",
		"\U0001d54b\u210d\U0001d53c \u211a\U0001d54c\U0001d540\u2102\U0001d542 \U0001d539\u211d\U0001d546\U0001d54e\u2115 \U0001d53d\U0001d546\U0001d54f \U0001d541\U0001d54c\U0001d544\u2119\U0001d54a \U0001d546\U0001d54d\U0001d53c\u211d \U0001d54b\u210d\U0001d53c \U0001d543\U0001d538\u2124\U0001d550 \U0001d53b\U0001d546\U0001d53e \U0001d7d9\U0001d7da\U0001d7db\U0001d7dc\U0001d7dd\U0001d7de\U0001d7df\U0001d7e0\U0001d7e1\U0001d7d8",
		"\U0001d413\U0001d407\U0001d404 \U0001d410\U0001d414\U0001d408\U0001d402\U0001d40a \U0001d401\U0001d411\U0001d40e\U0001d416\U0001d40d \U0001d405\U0001d40e\U0001d417 \U0001d409\U0001d414\U0001d40c\U0001d40f\U0001d412 \U0001d40e\U0001d415\U0001d404\U0001d411 \U0001d413\U0001d407\U0001d404 \U0001d40b\U0001d400\U0001d419\U0001d418 \U0001d403\U0001d40e\U0001d406 \U0001d7cf\U0001d7d0\U0001d7d1\U0001d7d2\U0001d7d3\U0001d7d4\U0001d7d5\U0001d7d6\U0001d7d7\U0001d7ce",
		"\U0001d47b\U0001d46f\U0001d46c \U0001d478\U0001d47c\U0001d470\U0001d46a\U0001d472 \U0001d469\U0001d479\U0001d476\U0001d47e\U0001d475 \U0001d46d\U0001d476\U0001d47f \U0001d471\U0001d47c\U0001d474\U0001d477\U0001d47a \U0001d476\U0001d47d\U0001d46c\U0001d479 \U0001d47b\U0001d46f\U0001d46c \U0001d473\U0001d468\U0001d481\U0001d480 \U0001d46b\U0001d476\U0001d46e 1234567890",
		"\U0001d4e3\U0001d4d7\U0001d4d4 \U0001d4e0\U0001d4e4\U0001d4d8\U0001d4d2\U0001d4da \U0001d4d1\U0001d4e1\U0001d4de\U0001d4e6\U0001d4dd \U0001d4d5\U0001d4de\U0001d4e7 \U0001d4d9\U0001d4e4\U0001d4dc\U0001d4df\U0001d4e2 \U0001d4de\U0001d4e5\U0001d4d4\U0001d4e1 \U0001d4e3\U0001d4d7\U0001d4d4 \U0001d4db\U0001d4d0\U0001d4e9\U0001d4e8 \U0001d4d3\U0001d4de\U0001d4d6 1234567890",
		"\U0001d57f\U0001d573\U0001d570 \U0001d57c\U0001d580\U0001d574\U0001d56e\U0001d576 \U0001d56d\U0001d57d\U0001d57a\U0001d582\U0001d579 \U0001d571\U0001d57a\U0001d583 \U0001d575\U0001d580\U0001d578\U0001d57b\U0001d57e \U0001d57a\U0001d581\U0001d570\U0001d57d \U0001d57f\U0001d573\U0001d570 \U0001d577\U0001d56c\U0001d585\U0001d584 \U0001d56f\U0001d57a\U0001d572 1234567890",
	}
	for _, c := range cases {
		testUnidecode(t, c, v)
	}
}

func TestUnidecodeConverterB(t *testing.T) {
	v := "the quick brown fox jumps over the lazy dog 1234567890"
	cases := []string{
		"\uff54\uff48\uff45 \uff51\uff55\uff49\uff43\uff4b \uff42\uff52\uff4f\uff57\uff4e \uff46\uff4f\uff58 \uff4a\uff55\uff4d\uff50\uff53 \uff4f\uff56\uff45\uff52 \uff54\uff48\uff45 \uff4c\uff41\uff5a\uff59 \uff44\uff4f\uff47 \uff11\uff12\uff13\uff14\uff15\uff16\uff17\uff18\uff19\uff10",
		"\U0001d565\U0001d559\U0001d556 \U0001d562\U0001d566\U0001d55a\U0001d554\U0001d55c \U0001d553\U0001d563\U0001d560\U0001d568\U0001d55f \U0001d557\U0001d560\U0001d569 \U0001d55b\U0001d566\U0001d55e\U0001d561\U0001d564 \U0001d560\U0001d567\U0001d556\U0001d563 \U0001d565\U0001d559\U0001d556 \U0001d55d\U0001d552\U0001d56b\U0001d56a \U0001d555\U0001d560\U0001d558 \U0001d7d9\U0001d7da\U0001d7db\U0001d7dc\U0001d7dd\U0001d7de\U0001d7df\U0001d7e0\U0001d7e1\U0001d7d8",
		"\U0001d42d\U0001d421\U0001d41e \U0001d42a\U0001d42e\U0001d422\U0001d41c\U0001d424 \U0001d41b\U0001d42b\U0001d428\U0001d430\U0001d427 \U0001d41f\U0001d428\U0001d431 \U0001d423\U0001d42e\U0001d426\U0001d429\U0001d42c \U0001d428\U0001d42f\U0001d41e\U0001d42b \U0001d42d\U0001d421\U0001d41e \U0001d425\U0001d41a\U0001d433\U0001d432 \U0001d41d\U0001d428\U0001d420 \U0001d7cf\U0001d7d0\U0001d7d1\U0001d7d2\U0001d7d3\U0001d7d4\U0001d7d5\U0001d7d6\U0001d7d7\U0001d7ce",
		"\U0001d495\U0001d489\U0001d486 \U0001d492\U0001d496\U0001d48a\U0001d484\U0001d48c \U0001d483\U0001d493\U0001d490\U0001d498\U0001d48f \U0001d487\U0001d490\U0001d499 \U0001d48b\U0001d496\U0001d48e\U0001d491\U0001d494 \U0001d490\U0001d497\U0001d486\U0001d493 \U0001d495\U0001d489\U0001d486 \U0001d48d\U0001d482\U0001d49b\U0001d49a \U0001d485\U0001d490\U0001d488 1234567890",
		"\U0001d4fd\U0001d4f1\U0001d4ee \U0001d4fa\U0001d4fe\U0001d4f2\U0001d4ec\U0001d4f4 \U0001d4eb\U0001d4fb\U0001d4f8\U0001d500\U0001d4f7 \U0001d4ef\U0001d4f8\U0001d501 \U0001d4f3\U0001d4fe\U0001d4f6\U0001d4f9\U0001d4fc \U0001d4f8\U0001d4ff\U0001d4ee\U0001d4fb \U0001d4fd\U0001d4f1\U0001d4ee \U0001d4f5\U0001d4ea\U0001d503\U0001d502 \U0001d4ed\U0001d4f8\U0001d4f0 1234567890",
		"\U0001d599\U0001d58d\U0001d58a \U0001d596\U0001d59a\U0001d58e\U0001d588\U0001d590 \U0001d587\U0001d597\U0001d594\U0001d59c\U0001d593 \U0001d58b\U0001d594\U0001d59d \U0001d58f\U0001d59a\U0001d592\U0001d595\U0001d598 \U0001d594\U0001d59b\U0001d58a\U0001d597 \U0001d599\U0001d58d\U0001d58a \U0001d591\U0001d586\U0001d59f\U0001d59e \U0001d589\U0001d594\U0001d58c 1234567890",
	}
	for _, c := range cases {
		testUnidecode(t, c, v)
	}
}


unidecode.go
package unidecode

import (
	"strings"
	"unicode"

	"github.com/lemontreeran/go-unidecode/table"
)

// Version return version
func Version() string {
	return "0.1.1"
}

// Unidecode implements transliterate Unicode text into plain 7-bit ASCII.
// e.g. Unidecode("kožušček") => "kozuscek"
func Unidecode(s string) string {
	return unidecode(s)
}

func unidecode(s string) string {
	ret := []string{}
	for _, r := range s {
		if r < unicode.MaxASCII {
			v := string(r)
			ret = append(ret, v)
			continue
		}
		if r > 0xeffff {
			continue
		}

		section := r >> 8   // Chop off the last two hex digits
		position := r % 256 // Last two hex digits
		if tb, ok := table.Tables[section]; ok {
			if len(tb) > int(position) {
				ret = append(ret, tb[position])
			}
		}
	}
	return strings.Join(ret, "")
}


table/x05e.go
package table

var x05e = []string{
	"Za ",     // 0x00
	"Bi ",     // 0x01
	"Shi ",    // 0x02
	"Bu ",     // 0x03
	"Ding ",   // 0x04
	"Shuai ",  // 0x05
	"Fan ",    // 0x06
	"Nie ",    // 0x07
	"Shi ",    // 0x08
	"Fen ",    // 0x09
	"Pa ",     // 0x0a
	"Zhi ",    // 0x0b
	"Xi ",     // 0x0c
	"Hu ",     // 0x0d
	"Dan ",    // 0x0e
	"Wei ",    // 0x0f
	"Zhang ",  // 0x10
	"Tang ",   // 0x11
	"Dai ",    // 0x12
	"Ma ",     // 0x13
	"Pei ",    // 0x14
	"Pa ",     // 0x15
	"Tie ",    // 0x16
	"Fu ",     // 0x17
	"Lian ",   // 0x18
	"Zhi ",    // 0x19
	"Zhou ",   // 0x1a
	"Bo ",     // 0x1b
	"Zhi ",    // 0x1c
	"Di ",     // 0x1d
	"Mo ",     // 0x1e
	"Yi ",     // 0x1f
	"Yi ",     // 0x20
	"Ping ",   // 0x21
	"Qia ",    // 0x22
	"Juan ",   // 0x23
	"Ru ",     // 0x24
	"Shuai ",  // 0x25
	"Dai ",    // 0x26
	"Zheng ",  // 0x27
	"Shui ",   // 0x28
	"Qiao ",   // 0x29
	"Zhen ",   // 0x2a
	"Shi ",    // 0x2b
	"Qun ",    // 0x2c
	"Xi ",     // 0x2d
	"Bang ",   // 0x2e
	"Dai ",    // 0x2f
	"Gui ",    // 0x30
	"Chou ",   // 0x31
	"Ping ",   // 0x32
	"Zhang ",  // 0x33
	"Sha ",    // 0x34
	"Wan ",    // 0x35
	"Dai ",    // 0x36
	"Wei ",    // 0x37
	"Chang ",  // 0x38
	"Sha ",    // 0x39
	"Qi ",     // 0x3a
	"Ze ",     // 0x3b
	"Guo ",    // 0x3c
	"Mao ",    // 0x3d
	"Du ",     // 0x3e
	"Hou ",    // 0x3f
	"Zheng ",  // 0x40
	"Xu ",     // 0x41
	"Mi ",     // 0x42
	"Wei ",    // 0x43
	"Wo ",     // 0x44
	"Fu ",     // 0x45
	"Yi ",     // 0x46
	"Bang ",   // 0x47
	"Ping ",   // 0x48
	"Tazuna ", // 0x49
	"Gong ",   // 0x4a
	"Pan ",    // 0x4b
	"Huang ",  // 0x4c
	"Dao ",    // 0x4d
	"Mi ",     // 0x4e
	"Jia ",    // 0x4f
	"Teng ",   // 0x50
	"Hui ",    // 0x51
	"Zhong ",  // 0x52
	"Shan ",   // 0x53
	"Man ",    // 0x54
	"Mu ",     // 0x55
	"Biao ",   // 0x56
	"Guo ",    // 0x57
	"Ze ",     // 0x58
	"Mu ",     // 0x59
	"Bang ",   // 0x5a
	"Zhang ",  // 0x5b
	"Jiong ",  // 0x5c
	"Chan ",   // 0x5d
	"Fu ",     // 0x5e
	"Zhi ",    // 0x5f
	"Hu ",     // 0x60
	"Fan ",    // 0x61
	"Chuang ", // 0x62
	"Bi ",     // 0x63
	"Hei ",    // 0x64
	"[?] ",    // 0x65
	"Mi ",     // 0x66
	"Qiao ",   // 0x67
	"Chan ",   // 0x68
	"Fen ",    // 0x69
	"Meng ",   // 0x6a
	"Bang ",   // 0x6b
	"Chou ",   // 0x6c
	"Mie ",    // 0x6d
	"Chu ",    // 0x6e
	"Jie ",    // 0x6f
	"Xian ",   // 0x70
	"Lan ",    // 0x71
	"Gan ",    // 0x72
	"Ping ",   // 0x73
	"Nian ",   // 0x74
	"Qian ",   // 0x75
	"Bing ",   // 0x76
	"Bing ",   // 0x77
	"Xing ",   // 0x78
	"Gan ",    // 0x79
	"Yao ",    // 0x7a
	"Huan ",   // 0x7b
	"You ",    // 0x7c
	"You ",    // 0x7d
	"Ji ",     // 0x7e
	"Yan ",    // 0x7f
	"Pi ",     // 0x80
	"Ting ",   // 0x81
	"Ze ",     // 0x82
	"Guang ",  // 0x83
	"Zhuang ", // 0x84
	"Mo ",     // 0x85
	"Qing ",   // 0x86
	"Bi ",     // 0x87
	"Qin ",    // 0x88
	"Dun ",    // 0x89
	"Chuang ", // 0x8a
	"Gui ",    // 0x8b
	"Ya ",     // 0x8c
	"Bai ",    // 0x8d
	"Jie ",    // 0x8e
	"Xu ",     // 0x8f
	"Lu ",     // 0x90
	"Wu ",     // 0x91
	"[?] ",    // 0x92
	"Ku ",     // 0x93
	"Ying ",   // 0x94
	"Di ",     // 0x95
	"Pao ",    // 0x96
	"Dian ",   // 0x97
	"Ya ",     // 0x98
	"Miao ",   // 0x99
	"Geng ",   // 0x9a
	"Ci ",     // 0x9b
	"Fu ",     // 0x9c
	"Tong ",   // 0x9d
	"Pang ",   // 0x9e
	"Fei ",    // 0x9f
	"Xiang ",  // 0xa0
	"Yi ",     // 0xa1
	"Zhi ",    // 0xa2
	"Tiao ",   // 0xa3
	"Zhi ",    // 0xa4
	"Xiu ",    // 0xa5
	"Du ",     // 0xa6
	"Zuo ",    // 0xa7
	"Xiao ",   // 0xa8
	"Tu ",     // 0xa9
	"Gui ",    // 0xaa
	"Ku ",     // 0xab
	"Pang ",   // 0xac
	"Ting ",   // 0xad
	"You ",    // 0xae
	"Bu ",     // 0xaf
	"Ding ",   // 0xb0
	"Cheng ",  // 0xb1
	"Lai ",    // 0xb2
	"Bei ",    // 0xb3
	"Ji ",     // 0xb4
	"An ",     // 0xb5
	"Shu ",    // 0xb6
	"Kang ",   // 0xb7
	"Yong ",   // 0xb8
	"Tuo ",    // 0xb9
	"Song ",   // 0xba
	"Shu ",    // 0xbb
	"Qing ",   // 0xbc
	"Yu ",     // 0xbd
	"Yu ",     // 0xbe
	"Miao ",   // 0xbf
	"Sou ",    // 0xc0
	"Ce ",     // 0xc1
	"Xiang ",  // 0xc2
	"Fei ",    // 0xc3
	"Jiu ",    // 0xc4
	"He ",     // 0xc5
	"Hui ",    // 0xc6
	"Liu ",    // 0xc7
	"Sha ",    // 0xc8
	"Lian ",   // 0xc9
	"Lang ",   // 0xca
	"Sou ",    // 0xcb
	"Jian ",   // 0xcc
	"Pou ",    // 0xcd
	"Qing ",   // 0xce
	"Jiu ",    // 0xcf
	"Jiu ",    // 0xd0
	"Qin ",    // 0xd1
	"Ao ",     // 0xd2
	"Kuo ",    // 0xd3
	"Lou ",    // 0xd4
	"Yin ",    // 0xd5
	"Liao ",   // 0xd6
	"Dai ",    // 0xd7
	"Lu ",     // 0xd8
	"Yi ",     // 0xd9
	"Chu ",    // 0xda
	"Chan ",   // 0xdb
	"Tu ",     // 0xdc
	"Si ",     // 0xdd
	"Xin ",    // 0xde
	"Miao ",   // 0xdf
	"Chang ",  // 0xe0
	"Wu ",     // 0xe1
	"Fei ",    // 0xe2
	"Guang ",  // 0xe3
	"Koc ",    // 0xe4
	"Kuai ",   // 0xe5
	"Bi ",     // 0xe6
	"Qiang ",  // 0xe7
	"Xie ",    // 0xe8
	"Lin ",    // 0xe9
	"Lin ",    // 0xea
	"Liao ",   // 0xeb
	"Lu ",     // 0xec
	"[?] ",    // 0xed
	"Ying ",   // 0xee
	"Xian ",   // 0xef
	"Ting ",   // 0xf0
	"Yong ",   // 0xf1
	"Li ",     // 0xf2
	"Ting ",   // 0xf3
	"Yin ",    // 0xf4
	"Xun ",    // 0xf5
	"Guang ",  // 0xf6
	"Ting ",   // 0xf7
	"Di ",     // 0xf8
	"Po ",     // 0xf9
	"Jian ",   // 0xfa
	"Hui ",    // 0xfb
	"Nai ",    // 0xfc
	"Hui ",    // 0xfd
	"Gong ",   // 0xfe
	"Nian ",   // 0xff
}


table/x060.go
package table

var x060 = []string{
	"Huai ",    // 0x00
	"Tai ",     // 0x01
	"Song ",    // 0x02
	"Wu ",      // 0x03
	"Ou ",      // 0x04
	"Chang ",   // 0x05
	"Chuang ",  // 0x06
	"Ju ",      // 0x07
	"Yi ",      // 0x08
	"Bao ",     // 0x09
	"Chao ",    // 0x0a
	"Min ",     // 0x0b
	"Pei ",     // 0x0c
	"Zuo ",     // 0x0d
	"Zen ",     // 0x0e
	"Yang ",    // 0x0f
	"Kou ",     // 0x10
	"Ban ",     // 0x11
	"Nu ",      // 0x12
	"Nao ",     // 0x13
	"Zheng ",   // 0x14
	"Pa ",      // 0x15
	"Bu ",      // 0x16
	"Tie ",     // 0x17
	"Gu ",      // 0x18
	"Hu ",      // 0x19
	"Ju ",      // 0x1a
	"Da ",      // 0x1b
	"Lian ",    // 0x1c
	"Si ",      // 0x1d
	"Chou ",    // 0x1e
	"Di ",      // 0x1f
	"Dai ",     // 0x20
	"Yi ",      // 0x21
	"Tu ",      // 0x22
	"You ",     // 0x23
	"Fu ",      // 0x24
	"Ji ",      // 0x25
	"Peng ",    // 0x26
	"Xing ",    // 0x27
	"Yuan ",    // 0x28
	"Ni ",      // 0x29
	"Guai ",    // 0x2a
	"Fu ",      // 0x2b
	"Xi ",      // 0x2c
	"Bi ",      // 0x2d
	"You ",     // 0x2e
	"Qie ",     // 0x2f
	"Xuan ",    // 0x30
	"Cong ",    // 0x31
	"Bing ",    // 0x32
	"Huang ",   // 0x33
	"Xu ",      // 0x34
	"Chu ",     // 0x35
	"Pi ",      // 0x36
	"Xi ",      // 0x37
	"Xi ",      // 0x38
	"Tan ",     // 0x39
	"Koraeru ", // 0x3a
	"Zong ",    // 0x3b
	"Dui ",     // 0x3c
	"[?] ",     // 0x3d
	"Ki ",      // 0x3e
	"Yi ",      // 0x3f
	"Chi ",     // 0x40
	"Ren ",     // 0x41
	"Xun ",     // 0x42
	"Shi ",     // 0x43
	"Xi ",      // 0x44
	"Lao ",     // 0x45
	"Heng ",    // 0x46
	"Kuang ",   // 0x47
	"Mu ",      // 0x48
	"Zhi ",     // 0x49
	"Xie ",     // 0x4a
	"Lian ",    // 0x4b
	"Tiao ",    // 0x4c
	"Huang ",   // 0x4d
	"Die ",     // 0x4e
	"Hao ",     // 0x4f
	"Kong ",    // 0x50
	"Gui ",     // 0x51
	"Heng ",    // 0x52
	"Xi ",      // 0x53
	"Xiao ",    // 0x54
	"Shu ",     // 0x55
	"S ",       // 0x56
	"Kua ",     // 0x57
	"Qiu ",     // 0x58
	"Yang ",    // 0x59
	"Hui ",     // 0x5a
	"Hui ",     // 0x5b
	"Chi ",     // 0x5c
	"Jia ",     // 0x5d
	"Yi ",      // 0x5e
	"Xiong ",   // 0x5f
	"Guai ",    // 0x60
	"Lin ",     // 0x61
	"Hui ",     // 0x62
	"Zi ",      // 0x63
	"Xu ",      // 0x64
	"Chi ",     // 0x65
	"Xiang ",   // 0x66
	"Nu ",      // 0x67
	"Hen ",     // 0x68
	"En ",      // 0x69
	"Ke ",      // 0x6a
	"Tong ",    // 0x6b
	"Tian ",    // 0x6c
	"Gong ",    // 0x6d
	"Quan ",    // 0x6e
	"Xi ",      // 0x6f
	"Qia ",     // 0x70
	"Yue ",     // 0x71
	"Peng ",    // 0x72
	"Ken ",     // 0x73
	"De ",      // 0x74
	"Hui ",     // 0x75
	"E ",       // 0x76
	"Kyuu ",    // 0x77
	"Tong ",    // 0x78
	"Yan ",     // 0x79
	"Kai ",     // 0x7a
	"Ce ",      // 0x7b
	"Nao ",     // 0x7c
	"Yun ",     // 0x7d
	"Mang ",    // 0x7e
	"Yong ",    // 0x7f
	"Yong ",    // 0x80
	"Yuan ",    // 0x81
	"Pi ",      // 0x82
	"Kun ",     // 0x83
	"Qiao ",    // 0x84
	"Yue ",     // 0x85
	"Yu ",      // 0x86
	"Yu ",      // 0x87
	"Jie ",     // 0x88
	"Xi ",      // 0x89
	"Zhe ",     // 0x8a
	"Lin ",     // 0x8b
	"Ti ",      // 0x8c
	"Han ",     // 0x8d
	"Hao ",     // 0x8e
	"Qie ",     // 0x8f
	"Ti ",      // 0x90
	"Bu ",      // 0x91
	"Yi ",      // 0x92
	"Qian ",    // 0x93
	"Hui ",     // 0x94
	"Xi ",      // 0x95
	"Bei ",     // 0x96
	"Man ",     // 0x97
	"Yi ",      // 0x98
	"Heng ",    // 0x99
	"Song ",    // 0x9a
	"Quan ",    // 0x9b
	"Cheng ",   // 0x9c
	"Hui ",     // 0x9d
	"Wu ",      // 0x9e
	"Wu ",      // 0x9f
	"You ",     // 0xa0
	"Li ",      // 0xa1
	"Jing ",    // 0xa2
	"Huan ",    // 0xa3
	"Cong ",    // 0xa4
	"Yi ",      // 0xa5
	"Yue ",     // 0xa6
	"Li ",      // 0xa7
	"Nin ",     // 0xa8
	"Nao ",     // 0xa9
	"E ",       // 0xaa
	"Que ",     // 0xab
	"Xuan ",    // 0xac
	"Qian ",    // 0xad
	"Wu ",      // 0xae
	"Min ",     // 0xaf
	"Cong ",    // 0xb0
	"Fei ",     // 0xb1
	"Bei ",     // 0xb2
	"Duo ",     // 0xb3
	"Cui ",     // 0xb4
	"Chang ",   // 0xb5
	"Men ",     // 0xb6
	"Li ",      // 0xb7
	"Ji ",      // 0xb8
	"Guan ",    // 0xb9
	"Guan ",    // 0xba
	"Xing ",    // 0xbb
	"Dao ",     // 0xbc
	"Qi ",      // 0xbd
	"Kong ",    // 0xbe
	"Tian ",    // 0xbf
	"Lun ",     // 0xc0
	"Xi ",      // 0xc1
	"Kan ",     // 0xc2
	"Kun ",     // 0xc3
	"Ni ",      // 0xc4
	"Qing ",    // 0xc5
	"Chou ",    // 0xc6
	"Dun ",     // 0xc7
	"Guo ",     // 0xc8
	"Chan ",    // 0xc9
	"Liang ",   // 0xca
	"Wan ",     // 0xcb
	"Yuan ",    // 0xcc
	"Jin ",     // 0xcd
	"Ji ",      // 0xce
	"Lin ",     // 0xcf
	"Yu ",      // 0xd0
	"Huo ",     // 0xd1
	"He ",      // 0xd2
	"Quan ",    // 0xd3
	"Tan ",     // 0xd4
	"Ti ",      // 0xd5
	"Ti ",      // 0xd6
	"Nie ",     // 0xd7
	"Wang ",    // 0xd8
	"Chuo ",    // 0xd9
	"Bu ",      // 0xda
	"Hun ",     // 0xdb
	"Xi ",      // 0xdc
	"Tang ",    // 0xdd
	"Xin ",     // 0xde
	"Wei ",     // 0xdf
	"Hui ",     // 0xe0
	"E ",       // 0xe1
	"Rui ",     // 0xe2
	"Zong ",    // 0xe3
	"Jian ",    // 0xe4
	"Yong ",    // 0xe5
	"Dian ",    // 0xe6
	"Ju ",      // 0xe7
	"Can ",     // 0xe8
	"Cheng ",   // 0xe9
	"De ",      // 0xea
	"Bei ",     // 0xeb
	"Qie ",     // 0xec
	"Can ",     // 0xed
	"Dan ",     // 0xee
	"Guan ",    // 0xef
	"Duo ",     // 0xf0
	"Nao ",     // 0xf1
	"Yun ",     // 0xf2
	"Xiang ",   // 0xf3
	"Zhui ",    // 0xf4
	"Die ",     // 0xf5
	"Huang ",   // 0xf6
	"Chun ",    // 0xf7
	"Qiong ",   // 0xf8
	"Re ",      // 0xf9
	"Xing ",    // 0xfa
	"Ce ",      // 0xfb
	"Bian ",    // 0xfc
	"Hun ",     // 0xfd
	"Zong ",    // 0xfe
	"Ti ",      // 0xff
}
